extensions-core/parquet-extensions/src/main/java/org/apache/druid/data/input/parquet/simple/ParquetGroupFlattenerMaker.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.parquet.schema.Type;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+import javax.annotation.Nullable;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class ParquetGroupFlattenerMaker implements ObjectFlatteners.FlattenerMaker<Group>
+{
+  private static final MappingProvider DEFAULT_MAPPING_PROVIDER = new MappingProvider()
+  {
+    @Override
+    public <T> T map(Object source, Class<T> targetType, Configuration configuration)
+    {
+      throw new NotImplementedException();
+    }
+
+    @Override
+    public <T> T map(Object source, TypeRef<T> targetType, Configuration configuration)
+    {
+      throw new NotImplementedException();
+    }
+  };
+
+  private final Configuration jsonPathConfiguration;
+  private final ParquetGroupConverter converter;
+
+  ParquetGroupFlattenerMaker(boolean binaryAsString)
+  {
+    this.converter = new ParquetGroupConverter(binaryAsString);
+    this.jsonPathConfiguration = Configuration.builder()
+                                              .jsonProvider(new ParquetGroupJsonProvider(converter))
+                                              .mappingProvider(DEFAULT_MAPPING_PROVIDER)
+                                              .options(EnumSet.of(Option.SUPPRESS_EXCEPTIONS))
+                                              .build();
+  }
+
+  @Override
+  public Set<String> discoverRootFields(Group obj)
+  {
+    return obj.getType()
+              .getFields()
+              .stream()
+              .filter(Type::isPrimitive)
+              .map(Type::getName)
+              .collect(Collectors.toSet());
+  }
+
+  @Override
+  public Object getRootField(Group obj, String key)
+  {
+    Object val = converter.convertField(obj, key);
+    return finalizeConversion(val);
+  }
+
+  @Override
+  public Function<Group, Object> makeJsonPathExtractor(String expr)
+  {
+    final JsonPath jsonPath = JsonPath.compile(expr);
+    return record -> {
+      Object val = jsonPath.read(record, jsonPathConfiguration);
+      return finalizeConversion(val);
+    };
+  }
+
+  @Nullable
+  @Override
+  public Function<Group, Object> makeJsonQueryExtractor(String expr)
+  {
+    return null;
+  }
+
+  /**
+   * After json conversion, wrapped list items can still need unwrapped. See
+   * {@link ParquetGroupConverter#isWrappedListPrimitive(Object)} and
+   * {@link ParquetGroupConverter#unwrapListPrimitive(Object)} for more details.
+   *
+   * @param o
+   *
+   * @return
+   */
+  private Object finalizeConversion(Object o)
+  {
+    // conversion can leave 'wrapped' list primitives
+    if (ParquetGroupConverter.isWrappedListPrimitive(o)) {
+      return converter.unwrapListPrimitive(o);
+    } else if (o instanceof List) {
+      List<Object> asList = (List<Object>) o;
+      if (asList.stream().allMatch(ParquetGroupConverter::isWrappedListPrimitive)) {
```

Comment:
Are there cases where such lists have some items "wrapped" but some items "unwrapped"? 

I'm wondering if the allMatch isWrappedListPrimitive check is necessary, is it possible and correct to remove that pass and just unwrap anything that's wrapped in the list?

Reply:
I don't know what is allowed to happen, but I think if the list is not homogeneous it is safer to not do this conversion rather than do a partial conversion. Leaving this as is for now